packages/editor/src/browser/editor-navigation-contribution.ts

```diff
@@ -0,0 +1,196 @@
+/*
+ * Copyright (C) 2018 TypeFox and others.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+import { inject, injectable, postConstruct } from 'inversify';
+import { ILogger } from '@theia/core/lib/common/logger';
+import { StorageService } from '@theia/core/lib/browser/storage-service';
+import { Disposable, DisposableCollection } from '@theia/core/lib/common/disposable';
+import { FrontendApplicationContribution } from '@theia/core/lib/browser/frontend-application';
+import { Command, CommandContribution, CommandRegistry } from '@theia/core/lib/common/command';
+import { KeybindingContribution, KeybindingRegistry } from '@theia/core/lib/browser/keybinding';
+import { MenuContribution, MenuModelRegistry, MAIN_MENU_BAR } from '@theia/core/lib/common/menu';
+import { EditorWidget } from './editor-widget';
+import { EditorManager } from './editor-manager';
+import { TextEditor, Position, Range, TextDocumentChangeEvent } from './editor';
+import { NavigationLocation } from './navigation/navigation-location';
+import { NavigationLocationService } from './navigation/navigation-location-service';
+
+@injectable()
+export class EditorNavigationContribution implements Disposable, MenuContribution, CommandContribution, KeybindingContribution, FrontendApplicationContribution {
+
+    private static ID = 'editor-navigation-contribution';
+
+    protected readonly toDispose = new DisposableCollection();
+    protected readonly toDisposePerCurrentEditor = new DisposableCollection();
+
+    @inject(ILogger)
+    protected readonly logger: ILogger;
+
+    @inject(EditorManager)
+    protected readonly editorManager: EditorManager;
+
+    @inject(NavigationLocationService)
+    protected readonly locationStack: NavigationLocationService;
+
+    @inject(StorageService)
+    protected readonly storageService: StorageService;
+
+    @postConstruct()
+    protected init(): void {
+        this.toDispose.pushAll([
+            // TODO listen on file resource changes, if a file gets deleted, remove the corresponding navigation locations (if any).
+            // This would require introducing the FS dependency in the editor extension.
+            this.editorManager.onCurrentEditorChanged(this.onCurrentEditorChanged.bind(this))
+        ]);
+    }
+
+    async onStart(): Promise<void> {
+        this.restoreState();
```

Comment:
`await`?

Reply:
👍 We can do that; I left it out intentionally.

Reply:
ok, i mostly was wondering whether it should have the same lifecycle as a layout. For example, in Atom there are hooks to store and restore data, we have it only for the layout. And Atom decides when it is a good time to store and restore. In our case it seems to be a responsibility of the extension developer. It could be bad in the long run for example if we change when the data should be stored on the shutdown then each extension should be updated. I will open an issue to discuss it separately.

Reply:
In this case, would it make sense to store the state only `onStop` similar how we store the layout data?

Reply:
Great idea, thanks! I was also wondering if we could do something like this in the `FrontendApplication#startContributions`:
```typescript
await Promise.all(this.contributions.getContributions().filter(c => !!c.onStart).map(c => c.onStart!(this)));
```
Instead of starting each contribution one by one and wait for them (if it is a Promise). We just need to handle errors properly, and not fail fast.

Reply:
> In this case, would it make sense to store the state only onStop similar how we store the layout data?

Sure, we can do that.

Reply:
> ...Instead of starting each contribution one by one and wait for them (if it is a Promise). We just need to handle errors properly, and not fail fast.

I am not sure that it will change anything since everything happens in the single thread. But in theory, yes it should be possible to rearrange the code like that. I won't do it now since there are assumptions on order of extensions loading and registration i cannot really tell whether it would break something.

Reply:
Alright, let's leave as is.

Reply:
> In this case, would it make sense to store the state only onStop similar how we store the layout data?

It raises a question; `FrontendApplicationContribution#onStop` says: ``` Note that this is implemented using `window.unload` which doesn't allow any asynchronous code anymore.```. Still, we call async code from `onStop`, for instance: `WorkspaceService`. Do we still want to go with this approach? It does not seem to be safe.

In a nutshell: `this.storageService.setData` is async and we should not call it from `window.unload`.

Reply:
We could have a sync API on the `StorageService`.

Reply:
> In a nutshell: this.storageService.setData is async and we should not call it from window.unload.

It is fine, it is implemented with `LocalStorage` which is synchronous. Maybe we should make `StorageService` synchronous as well. Let's keep the discussion in https://github.com/theia-ide/theia/issues/1671. It is not in the scope of this PR.

Reply:
OK.